bittide-instances/bin/Shake.hs

```diff
@@ -39,6 +39,63 @@ import qualified Clash.Util.Interpolate as I
 import qualified Language.Haskell.TH as TH
 import qualified System.Directory as Directory
 
+{- [Note: invalidating HDL cache]
+
+Ideally, Shake would be perfectly aware of all source files needed to build RTL
+generated by Clash. It would - among other things - read 'bittide-instances.cabal',
+discover dependencies, need _their_ cabal files, need the Haskell files they
+point to, etc. Any change in these files would then restart the recovery process,
+and instruct Clash to rebuild RTL.
+
+Alas, we use Cabal as a build tool hence Shake is unaware of the exact files
+needed to trigger a Clash build. As a workaround, we pessimistically assume that
+_any_ file in the repository (for details see 'getWatchFiles') can trigger a
+rebuild. Hence, the Clash build job will list:
+
+    needWatchFiles
+
+Unfortunately, this isn't quite enough still to trigger a rebuild when adding a
+new Haskell module. Even though Shake will discover an associated Cabal file to
+have changed†, but will not watch the new module added to that Cabal file. In
+order to work around this, 'needWatchFiles' is used twice:
+
+    needWatchFiles  -- will trigger on .cabal change
+    needWatchFiles  -- will discover new file
+
+In the future we might be more precise in our source discovery process.
+
+† Note that Cabal itself will misbehave when a file is not explicitly mentioned
+  in a cabal file, but is imported in a listed one. We therefore will not try to
+  handle this case in Shake either.
+-}
+
+-- | Get all files whose changes will trigger an HDL rebuild. Because we lack a
+-- reliable way to determine which files should trigger a rebuild, this function
+-- returns a (very) pessimistic list: all files in the project's directory,
+-- except files ignored by git and files matching patterns in 'ignorePatterns'.
+getWatchFiles :: IO [String]
+getWatchFiles = do
+  getWatchFilesPy <- getDataFileName ("data" </> "scripts" </> "get_watch_files.py")
+  lines <$> readProcess getWatchFilesPy ignorePatterns ""
+
+-- | Call 'need' on output of 'getWatchFiles'
+needWatchFiles :: Action ()
+needWatchFiles = need =<< liftIO getWatchFiles
+
+-- | File patterns of file we do _not_ want to make trigger a rebuild of HDL
+-- files. Also see 'getWatchFiles'.
+ignorePatterns :: [String]
+ignorePatterns =
+  [ "*.md"
+  , ".github"
+  , ".reuse"
+  , ".vscode"
+  , "LICENSES"
+
+  -- Used for synthesis, but not for generating Clash output:
+  , "bittide-instances/data/constraints/*.xdc"
+  ]
```

Comment:
Either add ".license" or remove "LICENSES". Since both rarely change, I prefer the latter.